splicemachine/spliceengine
# DB-11718 Refine index lookup cost estimation and evaluate constant string

## Short Description
This PR includes two changes:
* evaluate constant string to string conversion in binding
* index lookup cost tuning

## Long Description
In investigating query hash `SMopSK5W`, three problems are identified (for detailed explanation, see Jira). This PR addresses the first two issues.

1. So far, we don't evaluate a cast node if it's from string to string, for example, `cast('abc' as char(4))`. Predicate selectivity estimation is affected if a predicate contains such a cast node as one of its operand. Since it's not a constant node, we will go for default estimation rather than ask statistics for a concrete value.

The reason we don't evaluate this kind of cast might be that we could cast a constant string to a narrower type, leading to data truncation. The truncation itself is totally fine, but according to standard, we should raise a warning in this case and send this warning together with the final result set to client. This is something not really possible in binding phase. However, early evaluation should be possible for casts whose target types are wider.

2. Currently, our index lookup cost has a penalty of 4000 per row. There are two issues about this. Firstly, the penalty is too high compared to actual execution times. Secondly, it's an over simplification because index lookup happens in a batched manner and could have some parallelism. Batch size and the maximum number of batches that can run concurrently are controlled by settings in `SQLConfiguration`. This change tunes the index lookup cost to consider these parameters and lower lookup penalty.

## How to test
Note that this PR currently doesn't give a better plan for `SMopSK5W`. Problem 3 described in Jira is preventing the tuned plan to be chosen.

For change 1, casting a string literal to a not-narrower type should be evaluated right in binding phase. However, this difference is not observable on client side. See `StringTypesToStringTypesConversionIT` for more details about this.

For change 2, we should see IndexLookup plans be chosen even when the number of lookup rows are 5 digits. Based on experiments on cloud, looking up 65K rows are still fast, which is just 1.2x - 1.3x times slower than scanning but not looking up the same number of rows. See adaptions in `testDefaultSelectivityFactorHint` for this part. Now looking up 27K rows should be considered faster than doing full table scan in its original setup.
